### DISCUSSION TITLE
IFS STMF with ccsid 37

### DISCUSSION THREAD

**Discussion by ToshBimbra** [codefori/vscode-ibmi#480] [General]
Since there is no visual indicator shown in IFS explorer to indicate if a stmf has ccsid 37 or not then it is frustrating to open the stmf and find the stmf opens with illegible chars.
Not sure if this topic has been brought up before or not.
Any comments or thoughts about my post?
Thx. Tosh.

**Comment by worksofliam** [comment]
@ToshBimbra Good point. In general, VS Code only supports UTF-x, so any source other than UTF-x would need to be converted to UTF-8. We don't currently look at the source ccsid and convert as needed for streamfiles. We already do this for source members as you probably know.

I am curious: what is your use case with streamfiles in 37 instead of 1208?

@ThePrez what's the status of `getccsid`?

**Reply by ToshBimbra** [reply]
IFS has in many cases over taken QSYS lib system. 
I know Open Source has been well adopted for past several VRMs.
Although many utilities are (and should be) creating and using 819 or 1208 for ccsids there are still files that have ccsid 37.
I know there are log files that get generated in IFS with ccsid 37.
Opening these files in 5250 sessions is not for the faint of heart.  Many IFS files are likely to have record lengths greater then 132 columns.
This is where using client tooling to open IFS files has many benefits.

**Reply by worksofliam** [reply]
@ToshBimbra You got it. Thanks!

The next step is to raise an issue and I can get to work on this feature to read/handle different encodings in the IFS. If a Sponsors requires the functionality immediately, then I can work on it ASAP. Otherwise, it will go on my to-do list and likely be worked on next week. I am working on another big thing right now

**Reply by ToshBimbra** [reply]
Not anything urgent. You can priorities according to your workload. 
As you can guess, ACS provides me plenty of alternatives -)

**Reply by ThePrez** [reply]
> 
> @ThePrez what's the status of `getccsid`?

Thanks for the poke. I just need to clean it up a bit and publish. Hopefully I can make the time by this weekend

**Reply by ThePrez** [reply]
> 
> > 
> > @ThePrez what's the status of `getccsid`?
> 
> Thanks for the poke. I just need to clean it up a bit and publish. Hopefully I can make the time by this weekend
> 
> 

I can also add the capability to autoconvert the file to utf8 if you wish

**Reply by worksofliam** [reply]
Great thanks. Looks like we have a good use case for it!

**Reply by worksofliam** [reply]
@ThePrez That's up to you. It looks like I am going to need convert to and from UTF8 from many CCSIDs.

**Comment by bobcozzi** [comment]
@worksofliam do you just need an SQL Function that retrieves the CCSID for IFS files? I have one I can donate. But I suppose the bigger issue is reading code from the IFS in native CCSID, and converting it to UTF-8 and then reversing the process back. IBM SQL Table functions can automatically convert to the job CCSID or to a specific CCSID, such as 1208. They do now provide an IFS_READ interface (like I didn't in SQL Tools for a while) but saving it back in the target file's CCSID _might_ require some C/++ code on the IBM i side of things. (I assume you're using SQL to get/put the data from the IBM i.)

**Reply by worksofliam** [reply]
@bobcozzi This is already implemented and released. We make use of `/usr/bin/attr` to get the tagged CCSID and `/usr/bin/iconv` to convert streamfiles.
